manta-util/src/time/lock.rs

```diff
@@ -0,0 +1,196 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Time-synchronized and Time-locked Data
+
+use crate::time::{Duration, Instant};
+use core::{mem, ops::Deref};
+
+/// Timed Data
+#[derive(Copy, Debug, Eq, Hash, PartialEq)]
+pub struct Timed<T> {
+    /// Value
+    value: T,
+
+    /// Instant
+    instant: Instant,
+}
+
+impl<T> Timed<T> {
+    /// Builds a new [`Timed`] object over `value`.
+    #[inline]
+    pub fn new(value: T) -> Self {
+        Self::new_unchecked(value, Instant::now())
+    }
+
+    /// Builds a new [`Timed`] object over `value` created at the given `instant` without checking
+    /// that `instant` is [`Instant::now`].
+    #[inline]
+    pub const fn new_unchecked(value: T, instant: Instant) -> Self {
+        Self { instant, value }
+    }
+
+    /// Returns a shared reference to the underlying data.
+    #[inline]
+    pub const fn get(&self) -> &T {
+        &self.value
+    }
+
+    /// Returns the last [`Instant`] that `self` was modified. See [`elapsed`](Self::elapsed) to get
+    /// the amount of time since the last modification.
+    #[inline]
+    pub const fn modified_at(&self) -> Instant {
+        self.instant
+    }
+
+    /// Returns the amount of time that has elapsed since the last modification of the underlying
+    /// value. See [`modified_at`](Self::modified_at) to get the [`Instant`] of the last
+    /// modification.
+    #[inline]
+    pub fn elapsed(&self) -> Duration {
+        self.instant.elapsed()
+    }
+
+    /// Returns `true` if the amount of time elapsed since the last modification is larger than the
+    /// `timeout`.
+    #[inline]
+    pub fn has_expired(&self, timeout: Duration) -> bool {
+        self.elapsed() >= timeout
+    }
+
+    /// Resets the modification time to the value returned by [`Instant::now`].
+    #[inline]
+    pub fn tap(&mut self) {
+        self.instant = Instant::now();
+    }
+
+    /// Sets the internal value to `value` returning the old value.
+    #[inline]
+    pub fn set(&mut self, value: T) -> T {
+        self.mutate(move |t| mem::replace(t, value))
+    }
+
+    /// Mutates the internal value using `f`, resetting the modification time to [`Instant::now`].
+    #[inline]
+    pub fn mutate<F, R>(&mut self, f: F) -> R
+    where
+        F: FnOnce(&mut T) -> R,
+    {
+        self.mutate_timed(move |value, _| f(value))
+    }
+
+    /// Mutates the internal value with the [`Instant`] of the last modification to `self`,
+    /// resetting the modification time to [`Instant::now`].
+    #[inline]
+    pub fn mutate_timed<F, R>(&mut self, f: F) -> R
+    where
+        F: FnOnce(&mut T, Instant) -> R,
+    {
+        let result = f(&mut self.value, self.instant);
+        self.tap();
+        result
+    }
```

Comment:
Can we merge these two functions? Seems to not have significant difference.

Reply:
We don't want to use `mutate_timed` if we don't care about the creation time, so we shouldn't burden the users of the API with that requirement on their functions.